File: hw4/task3/Stack.cpp
#include <iostream>
#include "Stack.h"

using namespace std;

struct StackElement
{
    char data;
    StackElement *next;
};

struct Stack
{
    StackElement *top;
    int size;
};

Stack* createStack()
{
    Stack *s = new Stack;
    s->top = nullptr;
    s->size = 0;
    return s;
}

void deleteStack(Stack *s)
{
    if (isEmpty(s))
    {
        delete s;
        return;
    }

    while (s->size != 0)
    {
        pop(s);
    }
    delete s;
}

void push(Stack *s, char value)
{
    StackElement *newElement = new StackElement;
    newElement->data = value;
    newElement->next = s->top;
    s->top = newElement;
    s->size++;
}

bool isEmpty(Stack const *s)
{
    return s->size == 0;
}

char pop(Stack *s)
{
    if (isEmpty(s))
        return -1;

    char x = s->top->data;
    StackElement *toDelete = s->top;
    s->top = s->top->next;
    delete toDelete;
    s->size--;
    return x;
}

char peek(Stack *s)
{
    if (isEmpty(s))
        return -1;

    char x = s->top->data;
    return x;
}

File: sеm2/hw1/task1/src/sem2/hw1/task1/Stack.java
package sem2.hw1.task1;

public class Stack<Type> {
    private Element head;

    public Stack(){

    }

    public void push(Type value){
        head = new Element(value, head);
    }

    public Type pop(){
        if (isEmpty())
            return null;

        Type popped = head.value;
        head = head.next;
        return popped;
    }

    public boolean isEmpty(){
        return head == null;
    }

    private class Element {
        private Type value;
        private Element next;

        public Element(Type value, Element next){
            this.value = value;
            this.next = next;
        }
    }
}
